src/plonk_util.rs

```diff
@@ -77,6 +74,44 @@ pub(crate) fn halo_n<C: HaloCurve>(s_bits: &[bool]) -> C::ScalarField {
     a * C::ZETA_SCALAR + b
 }
 
+/// Compute `n(x)` for a given `x`, where `n` is the injective function related to the Halo
+/// endomorphism.
+pub(crate) fn halo_n_mul<C: HaloCurve>(s_bits: &[bool], p: AffinePoint<C>) -> AffinePoint<C> {
+    // This is based on Algorithm 1 of the Halo paper, except that we start with Acc = O.
+
+    debug_assert_eq!(s_bits.len() % 2, 0, "Number of scalar bits must be even");
+
+    let p_p = p.to_projective();
+    let p_n = p_p.neg();
+    let endo_p_p = C::endomorphism(p).to_projective();
+    let endo_p_n = endo_p_p.neg();
+
+    let mut acc = ProjectivePoint::<C>::ZERO;
+
+    for s_bits_chunk in s_bits.chunks(2) {
+        let bit_lo = s_bits_chunk[0];
+        let bit_hi = s_bits_chunk[1];
+
+        let s;
```

Comment:
nit: maybe slightly nicer as `let s = if bit_hi { ... } else { ... }`, or do you not like that style?